### PR TITLE
[MoE] Expose zero-copy MegaMoE workspace output view

### PR DESCRIPTION
**Pull request by foraxe**
## Summary

This draft exposes a backward-compatible zero-copy output path for the
FlashInfer MegaMoE layer.

The new `return_workspace_view=False` argument preserves the existing owned
output behavior by default. When enabled and supported by the selected
MegaMoE backend, `forward` returns the layer's workspace output view instead
of copying the result into a new tensor. Capability gating keeps unsupported
backends on the existing path. Backend support is expressed as the
`supports_output_view` capability property on the kernel contract.

The patch also keeps the workspace-view contract explicit across the NVFP4 and
MXFP8 CuTeDSL MegaMoE backends and updates the relevant CUDA-graph and
multi-rank tests.

## Before and after

```text
Before (default behavior):

  MoEEpTensors
       |
       v
  MegaMoE forward(...)
       |
       v
  workspace output --copy--> caller-owned output tensor

After (opt-in view):

  MoEEpTensors
       |
       v
  MegaMoE forward(..., return_workspace_view=True)
       |
       v
  workspace output view ------> caller uses workspace-backed tensor
```

The default remains the copied, caller-owned output. The new path makes the
workspace view available only when the backend advertises the capability.

## Motivation

The SGLang MegaMoE integration currently needs the computed output to remain
in the FlashInfer workspace. Returning that view removes a large per-layer
output materialization while retaining the old API behavior for callers that
need an owned tensor.

## Validation

- `tests/moe_ep/test_mega_cuda_graph.py`: 8 passed.
- The targeted output-view test passes: 1 passed, 7 deselected.
- Four-GPU NVFP4 multi-rank output-view coverage passes with
  `NVSHMEM_DISABLE_CUDA_VMM=1`.
- The corresponding SGLang adapter test passes 3 cases when paired with this
  API.

The upstream `moe_ep_benchmark` harness was also run from the
`vllm_repro_8_gpu_v2` branch with the FlashInfer MegaMoE section on four GPUs
(`GPUS=4`, `DEVS=0,1,2,3`). This is a GB200 smoke/regression run, not an
8-GPU reference result:

| backend | tokens/rank | hidden/intermediate | experts/top-k | p50 latency | throughput |
| --- | ---: | ---: | ---: | ---: | ---: |
| NVFP4 CuTeDSL | 32 | 7168 / 2048 | 256 / 8 | 358.4 us | 357174.7 |
| MXFP8 CuTeDSL | 32 | 7168 / 2048 | 256 / 8 | 619.9 us | 206472.9 |

The harness reported accuracy-loss fields of 23.201 percent for NVFP4 and
6.371 percent for MXFP8 against its BF16 dense reference. These are retained
as harness output and are not used as a model accuracy claim here.

This is the FlashInfer side of the paired SGLang integration. The SGLang
draft depends on this API and is posted separately.

## Summary by CodeRabbit

* **New Features**
  * Added an option to return workspace-backed output views, reducing unnecessary output copying when supported.
  * Added capability detection so unsupported configurations provide a clear error while preserving existing default behavior.
  * Enabled output views for supported FP8 and FP4 workflows.

* **Tests**
  * Added coverage for output-view shapes, workspace aliasing, repeatability, zero-token inputs, and multi-size execution.

### CODE REVIEW

File: flashinfer/moe_ep/backends/mega/kernel/mxfp8_cutedsl/backend.py [hunk 1]
```diff
@@ -206,6 +206,9 @@ def stage_inputs(
 
             note_staged_tokens(workspace.topk_idx, num_tokens)
 
+    def supports_output_view(self) -> bool:
+        return True
```
Comment (mhoqueanik) [inline flashinfer/moe_ep/backends/mega/kernel/mxfp8_cutedsl/backend.py]: can we streamline it? instead of a method, we could have a property in the contract that is set during new backend addition and the method reads from there?

File: flashinfer/moe_ep/backends/mega/kernel/nvfp4_cutedsl/backend.py [hunk 2]
```diff
@@ -210,6 +210,9 @@ def stage_inputs(
         if t.fc1_norm_const is not None:
             workspace.fc1_norm_const.copy_(t.fc1_norm_const)
 
+    def supports_output_view(self) -> bool:
+        return True
```
Comment (mhoqueanik) [inline flashinfer/moe_ep/backends/mega/kernel/nvfp4_cutedsl/backend.py]: ditto

File: flashinfer/moe_ep/core/kernel/base.py [hunk 3]
```diff
@@ -209,6 +209,10 @@ def stage_inputs(  # noqa: B027 - intentional no-op default
     ) -> None:
         """Copy or quantize activations into workspace buffers."""
 
+    def supports_output_view(self) -> bool:
+        """Whether ``compute(output=None)`` returns a workspace-backed view."""
+        return False
```
Comment (mhoqueanik) [inline flashinfer/moe_ep/core/kernel/base.py]: ditto

File: flashinfer/moe_ep/modes/mega_layer.py [hunk 4]
```diff
@@ -115,6 +115,11 @@ def _ensure_workspace(self) -> Any:
             )
         return self._workspace
 
+    @property
+    def supports_output_view(self) -> bool:
+        """Whether ``forward(return_workspace_view=True)`` is supported."""
+        return self._kernel.supports_output_view()
```
Comment (mhoqueanik) [inline flashinfer/moe_ep/modes/mega_layer.py]: this should work with the fix suggested above

**Review by mhoqueanik — COMMENTED** [review]
overall looks like a good suggestion. Can you briefly run the microbench from [here](https://github.com/mhoqueanik/moe_ep_benchmark/tree/vllm_repro_8_gpu_v2) to make sure the PR doesn't break the upstream paths?

File: flashinfer/moe_ep/backends/mega/kernel/mxfp8_cutedsl/backend.py (hunk 1 above)
Comment (foraxe) [reply flashinfer/moe_ep/backends/mega/kernel/mxfp8_cutedsl/backend.py]: Done — `supports_output_view` is now a capability property on the backend contract, defaulting to `False`, with the NVFP4 and MXFP8 backends opting in. `MegaLayer` now reads the property directly. I also ran the requested upstream microbenchmark and added the results to the PR description.

**Comment by foraxe** [comment]
> overall looks like a good suggestion. Can you briefly run the microbench from [here](https://github.com/mhoqueanik/moe_ep_benchmark/tree/vllm_repro_8_gpu_v2) to make sure the PR doesn't break the upstream paths?

@mhoqueanik  Yes — I ran the `vllm_repro_8_gpu_v2` harness on 4×GB200 (`GPUS=4`, `DEVS=0,1,2,3`) as a smoke/regression check.

Both upstream MegaMoE paths completed successfully:

* NVFP4 CuTeDSL: 358.4 µs p50, 357174.7 throughput
* MXFP8 CuTeDSL: 619.9 µs p50, 206472.9 throughput

I also reran the focused CUDA-graph, output-view, and 4-GPU multi-rank tests. The detailed results are included in the PR description.

This is not the original 8-GPU reference configuration, but I did not observe any failure or regression in the upstream paths.

**Comment by mhoqueanik** [comment]
Hi @foraxe, there is a minor conflict (it was introduced after the push style sm90 megamoe). Could you resolve it?

**Comment by foraxe** [comment]
Hi @mhoqueanik  Thanks for pointing this out; I rebased onto the latest main, resolved the SM90 MegaMoE conflict, and verified the relevant tests pass.

**Comment by mhoqueanik** [comment]
Hi @foraxe — while preparing the moe_ep merge queue we ran your PR merged with current main through the full moe_ep suite and hit one unit failure:

`tests/moe_ep/test_mega_layer_validation.py::test_mega_layer_allocates_output_before_staging_round`

Making the owned-output allocation conditional moved the `torch.empty` to after `stage_inputs()`, which breaks the allocate-before-staging contract that test enforces (from #4069 — allocator work between stage and compute can sync the device mid-round). It didn't surface in this PR's checks because CI hasn't actually run yet (it's gated on the `@flashinfer-bot run` authorization).

The fix keeps your conditional and just restores the ordering — we validated exactly this patch (unit + Blackwell mega multirank green on 8×B200):

```diff
--- a/flashinfer/moe_ep/modes/mega_layer.py
+++ b/flashinfer/moe_ep/modes/mega_layer.py
@@ -220,12 +220,9 @@ class MoEEpMegaLayer(nn.Module):
 
         workspace = self._ensure_workspace()
 
-        self._kernel.stage_inputs(
-            t,
-            workspace,
-            quantize_input=quantize_input,
-        )
-
+        # Owned-output allocation must stay ahead of the staging round (see
+        # test_mega_layer_allocates_output_before_staging_round): allocator
+        # work between stage and compute can sync the device mid-round.
         y = None
         if not return_workspace_view:
             y = torch.empty(
@@ -234,6 +231,13 @@ class MoEEpMegaLayer(nn.Module):
                 dtype=torch.bfloat16,
                 device=t.hidden_states.device,
             )
+
+        self._kernel.stage_inputs(
+            t,
+            workspace,
+            quantize_input=quantize_input,
+        )
+
         return self._kernel.compute(
             workspace,
             self._transformed,
```

No rebase needed — the branch merges cleanly with main; just this one commit, then a fresh `@flashinfer-bot run`. Also FYI: the incoming SM100 BF16 backend (#4386, merging right after this) intentionally does not opt into `supports_output_view` yet, so no action needed on your side for it.

**Comment by foraxe** [comment]
@flashinfer-bot run

**Comment by foraxe** [comment]
Thanks @mhoqueanik, I've applied the fix in https://github.com/flashinfer-ai/flashinfer/commit/982cb6aa1863c60930c9a1b08aefcaa21c3ffabd by moving output allocation before stage_inputs to satisfy the allocate-before-staging contract;
I also verified locally that test_mega_layer_allocates_output_before_staging_round is now passing.
It seems @flashinfer-bot run will need authorization.

**Comment by mhoqueanik** [comment]
@flashinfer-bot run

**Review by mhoqueanik — APPROVED** [review]

**Comment by foraxe** [comment]
We reran the unit tests in local env. (on GB200, Driver Version: 580.105.08, CUDA Version: 13.1)

  Ran with:
```
  CUDA_VISIBLE_DEVICES=0 FLASHINFER_DISABLE_VERSION_CHECK=1 /workspace/sglang_flashinfer/work/.venv-main/bin/python -m pytest -q <each file> --maxfail=1
```

  Local results:

  - test_mega_cuda_graph.py: 8 passed
  - test_fused_quant_stage.py: 11 passed
  - test_nvfp4_cutedsl_kernel_vs_reference.py: 4 passed
  - test_compute_bridge.py: 8 passed
  - test_deep_gemm_mega_kernel_vs_reference.py: 1 passed
  - test_mxfp8_cutedsl_preprocess_vs_reference.py: 3 passed
  - test_workspace_pool.py: 8 passed
  - test_moe_ep_nvfp4_cutedsl_mega_multirank.py: 7 passed, 14 skipped

  So in this local env, those DSLRuntimeError failures do not reproduce. They look environment/branch-variant specific to the CI stack (B300/GB200 + CUDA 13.0), not currently reproducible on this local setup. 

We also tested in https://github.com/sgl-project/sglang/pull/31470#issuecomment-5220175585

The new failures in bot pipeline seems like a broad MoE cutlass/DSL compiler regression on CUDA 13.0 + B300/GB200 (same DSLRuntimeError across many unrelated tests), not a narrow PR regression from this output-view/output-order fix.

**Comment by mhoqueanik** [comment]
@foraxe Agreed! I am looking into this
